### PR TITLE
GUACAMOLE-374: Generalize Docker image to automatically map environment variables to properties.

### PR DESCRIPTION
**Pull request by mike-jumper**
This change strips away much of the manual mapping of environment variables to properties, instead relying on automatic mapping of properties using `enable-environment-properties`. `*_FILE` environment variables are also automatically mapped.

Additionally:

* The structure of the startup process of the Docker image has been split across multiple scripts for maintainability and extensibility.
* The log level can now be set with a convenient `log-level` property that maps to the `LOG_LEVEL` environment variable.
* `REMOTE_IP_VALVE_*` environment variables have been renamed to align with the attributes being configured.
* The default `GUACAMOLE_HOME` is now transient and randomly generated to allow users to use the familiar `/etc/guacamole` location for their configuration template without potential collisions. Users can even use `/etc/guacamole` directly without overriding `GUACAMOLE_HOME` and rely on the image to combine what they've provided with anything the image has configured via environment variables.

As of these changes, there are only two mappings that must be manually maintained (and then only when a new extension is introduced):

1. A mapping of environment variable prefixes to their corresponding extension `.jar` files. This is handled by `guacamole-docker/build.d/010-map-guacamole-extensions.sh`.
2. A mapping of environment variable prefixes to their corresponding driver `.jar` files (ie: JDBC drivers). This is handled by `guacamole-docker/build.d/020-download-drivers.sh`.

Both of the above mappings/scripts organize things within a directory hierarchy that allows the entrypoint to easily link extensions, drivers, etc. into place based on which environment variables are set. Once the extensions are in place, everything else is implicit via `enable-environment-properties`.

In cases where environment variables are changing names, compatibility with the old naming is maintained with `guacamole-docker/entrypoint.d/000-migrate-legacy-variables.sh`. Old variables are automatically migrated to the new ones and a warning regarding deprecation is logged.

### CODE REVIEW

**Comment by mike-jumper** [comment]
I _believe_ this rewrite maintains parity with the features of the existing image, except that fewer sanity checks are performed. This revised process instead relies on the webapp itself to appropriately log errors related to configuration.

File: Dockerfile [hunk 1]
```diff
@@ -91,6 +96,11 @@ RUN useradd --system --create-home --shell /usr/sbin/nologin --uid $UID --gid $G
 # Run with user guacamole
 USER guacamole
 
+# Environment variable defaults
+ENV BAN_ENABLED=true \
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply Dockerfile:100]: This is actually effectively the same behavior that was originally introduced when [GUACAMOLE-990](https://issues.apache.org/jira/browse/GUACAMOLE-990) was originally implemented:

https://github.com/apache/guacamole-client/blob/83111616e531810262efdced5d26ed52ec630658/guacamole-docker/bin/start.sh#L1234-L1239

I don't recall whether we had a specific rationale for enabling this by default within the Docker image except that it's nearly always desirable.

File: guacamole-docker/build.d/000-build-and-install-guacamole.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+cd "$BUILD_DIR"
+mvn $MAVEN_ARGUMENTS package
+
+#
+# Copy built web application (guacamole.war) to destination location
+#
+
+mkdir -p "$DESTINATION/webapp"
+cp guacamole/target/*.war "$DESTINATION/webapp/guacamole.war"
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole-docker/build.d/000-build-and-install-guacamole.sh:49]: This is now accomplished via the following:

https://github.com/apache/guacamole-client/blob/c16ef1513bbbc46aa8a9f11fa79155195e35adac/guacamole-docker/entrypoint.d/500-generate-tomcat-catalina-base.sh#L48-L49

The environment variable for configuring this (`WEBAPP_CONTEXT`) remains unchanged.

[the remark this answers is not in the document]
Comment (necouchman) [reply guacamole-docker/build.d/000-build-and-install-guacamole.sh:49]: Ah, oops, sorry about that - hadn't made it quite that far in the review and made a bad assumption!

**Review by necouchman — CHANGES_REQUESTED** [review]
A couple more minor things, but looks pretty good otherwise.

@jmuehlner Anything from you on this one?

File: guacamole-docker/entrypoint.d/700-configure-features.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+##     Zero if the feature associated with the given environment variable
+##     prefix is enabled, non-zero otherwise.
+##
+is_feature_enabled() {
+
+    local VAR_BASE="$1"
+
+    # Allow any feature to be explicitly enabled/disabled using a
+    # [PREFIX]_ENABLED variable
+    local ENABLED_VAR="${VAR_BASE}ENABLED"
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole-docker/entrypoint.d/700-configure-features.sh:53]: Sure - I think I can clarify this.

File: guacamole-docker/entrypoint.d/700-configure-features.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+# under the License.
+#
+#
+
+##
+## @fn 800-configure-features.sh
+##
+## Automatically checks all environment variables currently set and performs
+## configuration tasks related to those variabels, including installing any
+## extensions ane external libraries associated with those variables to
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole-docker/entrypoint.d/700-configure-features.sh]: Oops. Fixed via rebase.

File: guacamole-docker/entrypoint.d/700-configure-features.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (mike-jumper) [reply guacamole-docker/entrypoint.d/700-configure-features.sh]: Fixed via rebase.

**Review by necouchman — APPROVED** [review]

**Comment by jmuehlner** [comment]
> A couple more minor things, but looks pretty good otherwise.
> 
> @jmuehlner Anything from you on this one?

Nope, seems like a reasonable change. If you're good with it, so am I.

**Comment by altmind** [comment]
its unclear how to customize GUACAMOLE_HOME now. regardless of the value in docker-compose environment setting, it defaults to /home/guacamole/.guacamole

is it because /home/guacamole/.guacamole path is present by default in docker image?

**Comment by mike-jumper** [comment]
@altmind Please don't use closed pull requests or JIRA issues to ask questions about Guacamole, even if you feel your question is relevant. The place to ask such questions would be the user@guacamole.apache.org mailing list:

https://guacamole.apache.org/support/#mailing-lists

Be sure to subscribe first or your email won't make it to the list.

**Pull request merged**